File: rust/crates/corlinman-gateway/src/py_config.rs
//! Rust → Python config handshake (Feature C last-mile).
//!
//! The Python side reads its `ProviderRegistry` + alias + embedding config
//! from a JSON file whose path is passed via the `CORLINMAN_PY_CONFIG` env
//! var. This module owns serialising the Rust [`Config`] into that JSON
//! shape (matching `corlinman_providers.specs.ProviderSpec` +
//! `AliasEntry` + `EmbeddingSpec`) and writing it atomically.
//!
//! The gateway calls [`write_py_config`] once at boot (so the Python
//! subprocess started by any supervisor sees a non-empty registry) and then
//! again after every admin write that mutates providers / aliases / models
//! / the embedding section.
//!
//! The JSON shape is:
//!
//! ```json
//! {
//!   "providers": [
//!     { "name": "anthropic", "kind": "anthropic",
//!       "api_key": "...", "base_url": null,
//!       "enabled": true, "params": {} }
//!   ],
//!   "aliases": {
//!     "smart": { "provider": "anthropic",
//!                "model": "claude-opus-4-7",
//!                "params": {"temperature": 0.7} }
//!   },
//!   "embedding": {
//!     "provider": "openai", "model": "text-embedding-3-small",
//!     "dimension": 1536, "enabled": true, "params": {}
//!   }
//! }
//! ```
//!
//! Secrets: `api_key` is resolved through [`SecretRef::resolve`] so the
//! JSON carries the concrete key string. This file is written with owner-
//! only permissions by virtue of living under `$CORLINMAN_DATA_DIR` (the
//! same directory as `config.toml`), and is overwritten on every admin
//! change.

use std::path::{Path, PathBuf};

use corlinman_core::config::{AliasEntry, Config, ProviderEntry, SecretRef};
use serde::Serialize;
use serde_json::{json, Value as JsonValue};

/// Env var name the Python side reads to locate the JSON drop.
pub const ENV_PY_CONFIG: &str = "CORLINMAN_PY_CONFIG";

/// Filename under `$CORLINMAN_DATA_DIR`.
const PY_CONFIG_FILENAME: &str = "py-config.json";

/// Default JSON path: `$CORLINMAN_DATA_DIR/py-config.json`, else
/// `/tmp/corlinman-py-config.json`. Mirrors the `/tmp`-friendly fallback
/// the Python side documents for container deployments.
pub fn default_py_config_path() -> PathBuf {
    if let Ok(dir) = std::env::var("CORLINMAN_DATA_DIR") {
        return PathBuf::from(dir).join(PY_CONFIG_FILENAME);
    }
    if let Some(home) = dirs::home_dir() {
        return home.join(".corlinman").join(PY_CONFIG_FILENAME);
    }
    PathBuf::from("/tmp/corlinman-py-config.json")
}

/// Provider spec matching `corlinman_providers.specs.ProviderSpec`.
#[derive(Debug, Serialize)]
struct PyProviderSpec {
    name: String,
    kind: &'static str,
    #[serde(skip_serializing_if = "Option::is_none")]
    api_key: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    base_url: Option<String>,
    enabled: bool,
    params: JsonValue,
}

/// Alias entry matching `corlinman_providers.specs.AliasEntry`.
#[derive(Debug, Serialize)]
struct PyAliasEntry {
    provider: String,
    model: String,
    params: JsonValue,
}

/// Embedding spec matching `corlinman_providers.specs.EmbeddingSpec`.
#[derive(Debug, Serialize)]
struct PyEmbeddingSpec {
    provider: String,
    model: String,
    dimension: u32,
    enabled: bool,
    params: JsonValue,
}

/// Render the Rust [`Config`] as the Python JSON shape.
///
/// Invariants:
/// * Providers without a resolvable `kind` are dropped (legacy slots with
///   neither explicit `kind` nor an inferred name — should never happen in
///   practice because `from_slot_name` covers every first-party name).
/// * Providers whose `api_key.resolve()` fails (env var unset) still
///   serialise with `api_key: null`, and Python treats that as "no auth"
///   (valid for local gateways) — we don't block config write just because
///   a dev forgot to export the key. Caller logs can mention the failure.
/// * Shorthand aliases (`smart = "claude-opus-4-7"`) become alias entries
///   with an *empty* `provider` field, because the Python `AliasEntry`
///   model requires a non-optional `provider`. Since the Python side also
///   supports legacy-prefix resolution, we omit shorthand aliases entirely
///   — the Python legacy-prefix fallback will route raw model ids like
///   `claude-*` to the right adapter on its own.
/// * Full-form aliases without an explicit `provider` are likewise omitted
///   — same reasoning.
pub fn render_py_config(cfg: &Config) -> JsonValue {
    let mut providers: Vec<PyProviderSpec> = Vec::new();
    for (name, entry) in cfg.providers.iter() {
        let Some(kind) = cfg.providers.kind_for(name, entry) else {
            continue;
        };
        providers.push(PyProviderSpec {
            name: name.to_string(),
            kind: kind.as_str(),
            api_key: resolve_api_key(entry),
            base_url: entry.base_url.clone(),
            enabled: entry.enabled,
            params: params_to_json(&entry.params),
        });
    }

    let mut aliases = serde_json::Map::new();
    for (alias_name, alias_entry) in cfg.models.aliases.iter() {
        // Only full-form aliases with an explicit provider are representable
        // as a Python AliasEntry. Shorthand / provider-less entries go
        // through the Python legacy-prefix fallback.
        let (provider, model, params) = match alias_entry {
            AliasEntry::Full(spec) => {
                let Some(p) = spec.provider.as_ref() else {
                    continue;
                };
                (p.clone(), spec.model.clone(), params_to_json(&spec.params))
            }
            AliasEntry::Shorthand(_) => continue,
        };
        aliases.insert(
            alias_name.clone(),
            serde_json::to_value(PyAliasEntry {
                provider,
                model,
                params,
            })
            .unwrap_or(JsonValue::Null),
        );
    }

    let embedding = cfg.embedding.as_ref().map(|emb| PyEmbeddingSpec {
        provider: emb.provider.clone(),
        model: emb.model.clone(),
        dimension: emb.dimension,
        enabled: emb.enabled,
        params: params_to_json(&emb.params),
    });

    json!({
        "providers": providers,
        "aliases": JsonValue::Object(aliases),
        "embedding": embedding,
    })
}

/// Write the Python-side JSON to `path` atomically (tmp + rename).
///
/// Creates parent dirs as needed. Callers should invoke this at boot
/// (after config load) and after every admin write that mutates the
/// relevant sections.
pub async fn write_py_config(cfg: &Config, path: &Path) -> std::io::Result<()> {
    let payload = render_py_config(cfg);
    let bytes = serde_json::to_vec_pretty(&payload).map_err(std::io::Error::other)?;
    if let Some(parent) = path.parent() {
        tokio::fs::create_dir_all(parent).await?;
    }
    let mut tmp = path.to_path_buf();
    tmp.as_mut_os_string().push(".new");
    tokio::fs::write(&tmp, &bytes).await?;
    tokio::fs::rename(&tmp, path).await?;
    Ok(())
}

/// Synchronous boot-path variant — used before any tokio runtime is
/// available (e.g. inside `fn build_runtime_with_logs` which is already
/// async but the write needs to happen before `set_var` and the Python
/// subprocess spawn). Having a sync version keeps the call sites shorter.
pub fn write_py_config_sync(cfg: &Config, path: &Path) -> std::io::Result<()> {
    let payload = render_py_config(cfg);
    let bytes = serde_json::to_vec_pretty(&payload).map_err(std::io::Error::other)?;
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    let mut tmp = path.to_path_buf();
    tmp.as_mut_os_string().push(".new");
    std::fs::write(&tmp, &bytes)?;
    std::fs::rename(&tmp, path)?;
    Ok(())
}

fn resolve_api_key(entry: &ProviderEntry) -> Option<String> {
    match entry.api_key.as_ref()? {
        SecretRef::Literal { value } => Some(value.clone()),
        SecretRef::EnvVar { env } => std::env::var(env).ok(),
    }
}

fn params_to_json(params: &corlinman_core::config::ParamsMap) -> JsonValue {
    // ParamsMap is `BTreeMap<String, serde_json::Value>`; direct serialise.
    serde_json::to_value(params).unwrap_or_else(|_| json!({}))
}

#[cfg(test)]
mod tests {
    use super::*;
    use corlinman_core::config::{
        AliasEntry, AliasSpec, Config, EmbeddingConfig, ParamsMap, ProviderEntry, ProviderKind,
        SecretRef,
    };

    fn cfg_with_everything() -> Config {
        let mut cfg = Config::default();
        std::env::set_var("PY_CONFIG_TEST_KEY", "sk-test-xyz");
        cfg.providers.anthropic = Some(ProviderEntry {
            kind: None, // inferred from slot name
            api_key: Some(SecretRef::EnvVar {
                env: "PY_CONFIG_TEST_KEY".into(),
            }),
            base_url: None,
            enabled: true,
            params: {
                let mut p = ParamsMap::new();
                p.insert("temperature".into(), serde_json::json!(0.7));
                p
            },
        });
        cfg.providers.openai = Some(ProviderEntry {
            kind: Some(ProviderKind::Openai),
            api_key: Some(SecretRef::Literal {
                value: "sk-literal".into(),
            }),
            base_url: Some("https://api.openai.com/v1".into()),
            enabled: true,
            params: ParamsMap::new(),
        });
        cfg.models.aliases.insert(
            "smart".into(),
            AliasEntry::Full(AliasSpec {
                model: "claude-opus-4-7".into(),
                provider: Some("anthropic".into()),
                params: {
                    let mut p = ParamsMap::new();
                    p.insert("temperature".into(), serde_json::json!(0.5));
                    p
                },
            }),
        );
        // Shorthand alias — should be omitted from JSON.
        cfg.models
            .aliases
            .insert("bare".into(), AliasEntry::Shorthand("gpt-4o".into()));
        cfg.embedding = Some(EmbeddingConfig {
            provider: "openai".into(),
            model: "text-embedding-3-small".into(),
            dimension: 1536,
            enabled: true,
            params: ParamsMap::new(),
        });
        cfg
    }

    #[test]
    fn render_matches_python_schema() {
        let cfg = cfg_with_everything();
        let v = render_py_config(&cfg);

        let providers = v["providers"].as_array().unwrap();
        assert_eq!(providers.len(), 2);
        // Anthropic — kind inferred, api_key resolved from env.
        let anthropic = providers.iter().find(|p| p["name"] == "anthropic").unwrap();
        assert_eq!(anthropic["kind"], "anthropic");
        assert_eq!(anthropic["api_key"], "sk-test-xyz");
        assert_eq!(anthropic["enabled"], true);
        assert_eq!(anthropic["params"]["temperature"], 0.7);
        // OpenAI — literal key, explicit kind + base_url.
        let openai = providers.iter().find(|p| p["name"] == "openai").unwrap();
        assert_eq!(openai["kind"], "openai");
        assert_eq!(openai["api_key"], "sk-literal");
        assert_eq!(openai["base_url"], "https://api.openai.com/v1");

        let aliases = v["aliases"].as_object().unwrap();
        // Only the full-form alias made it in.
        assert!(aliases.contains_key("smart"));
        assert!(!aliases.contains_key("bare"));
        assert_eq!(aliases["smart"]["provider"], "anthropic");
        assert_eq!(aliases["smart"]["model"], "claude-opus-4-7");
        assert_eq!(aliases["smart"]["params"]["temperature"], 0.5);

        let embedding = &v["embedding"];
        assert_eq!(embedding["provider"], "openai");
        assert_eq!(embedding["model"], "text-embedding-3-small");
        assert_eq!(embedding["dimension"], 1536);
        assert_eq!(embedding["enabled"], true);

        std::env::remove_var("PY_CONFIG_TEST_KEY");
    }

    #[test]
    fn write_py_config_sync_produces_parseable_file() {
        let tmp = tempfile::TempDir::new().unwrap();
        let path = tmp.path().join("py-config.json");
        let cfg = cfg_with_everything();

        write_py_config_sync(&cfg, &path).expect("write ok");
        let text = std::fs::read_to_string(&path).unwrap();
        let parsed: JsonValue = serde_json::from_str(&text).unwrap();
        assert!(parsed["providers"].is_array());
        assert!(parsed["aliases"].is_object());
        assert!(parsed["embedding"].is_object());
        // No stale .new sidecar.
        let mut stale = path.to_path_buf();
        stale.as_mut_os_string().push(".new");
        assert!(!stale.exists());

        std::env::remove_var("PY_CONFIG_TEST_KEY");
    }

    #[test]
    fn missing_env_var_leaves_api_key_null() {
        let mut cfg = Config::default();
        std::env::remove_var("PY_CONFIG_TEST_MISSING");
        cfg.providers.anthropic = Some(ProviderEntry {
            kind: None,
            api_key: Some(SecretRef::EnvVar {
                env: "PY_CONFIG_TEST_MISSING".into(),
            }),
            base_url: None,
            enabled: true,
            params: ParamsMap::new(),
        });
        let v = render_py_config(&cfg);
        let anthropic = &v["providers"][0];
        assert_eq!(anthropic["name"], "anthropic");
        assert!(
            anthropic["api_key"].is_null() || anthropic.get("api_key").is_none(),
            "unset env should render as null/absent, got {anthropic}"
        );
    }

    #[test]
    fn empty_config_renders_empty_sections() {
        let cfg = Config::default();
        let v = render_py_config(&cfg);
        assert!(v["providers"].as_array().unwrap().is_empty());
        assert!(v["aliases"].as_object().unwrap().is_empty());
        assert!(v["embedding"].is_null());
    }
}

File: python/packages/corlinman-server/src/corlinman_server/gateway/lifecycle/py_config.py
"""Rust→Python config handshake (Feature C last-mile).

Python port of ``rust/crates/corlinman-gateway/src/py_config.rs``.

Once Python *is* the runtime the Rust→Python file drop is partially
redundant — the same process that owns the live config can simply pass
:class:`ProviderRegistry` / alias maps into the agent servicer directly.
We keep the renderer / writer pair anyway, for three reasons:

1. ``corlinman_server.main._load_config`` still reads the
   ``CORLINMAN_PY_CONFIG`` JSON file when it's set — keeping the writer
   in-tree means a Python admin route that mutates ``providers`` /
   ``aliases`` / ``embedding`` can re-emit the file the existing
   in-process resolver watches.
2. External integrations (sidecars, FFI loaders) that grew up against
   the JSON shape stay supported with zero behavioural change.
3. The schema is small and the test suite anchors it — re-implementing
   it later if a sibling regresses the shape would be needless churn.

The JSON shape mirrors the Rust ``render_py_config`` byte-for-byte:

.. code-block:: json

    {
      "providers": [
        { "name": "anthropic", "kind": "anthropic",
          "api_key": "...", "base_url": null,
          "enabled": true, "params": {} }
      ],
      "aliases": {
        "smart": { "provider": "anthropic",
                   "model": "claude-opus-4-7",
                   "params": {"temperature": 0.7} }
      },
      "embedding": {
        "provider": "openai", "model": "text-embedding-3-small",
        "dimension": 1536, "enabled": true, "params": {}
      }
    }

The Python config object can be either a :class:`pydantic.BaseModel`,
a :class:`dict`, or any object whose ``providers`` / ``models.aliases``
/ ``embedding`` attributes behave like the Rust ``Config`` does. The
renderer is duck-typed (``getattr`` + ``hasattr``) so a future config
schema rev doesn't break the handshake.
"""

from __future__ import annotations

import json
import os
import tempfile
from collections.abc import Iterable, Mapping
from pathlib import Path
from typing import Any, Final

#: Env var name the Python AI plane reads to locate the JSON drop.
#: Mirrors ``corlinman_gateway::py_config::ENV_PY_CONFIG``.
ENV_PY_CONFIG: Final[str] = "CORLINMAN_PY_CONFIG"

#: Filename under ``$CORLINMAN_DATA_DIR``.
DEFAULT_PY_CONFIG_FILENAME: Final[str] = "py-config.json"

#: Per-spec the Python ``ProviderSpec.kind`` enum (mirrors
#: ``corlinman_core::config::ProviderKind``). Used to validate provider
#: ``kind`` values when the input config carries an explicit kind.
KNOWN_PROVIDER_KINDS: frozenset[str] = frozenset(
    {
        "anthropic",
        "openai",
        "openai_compatible",
        "gemini",
        "ollama",
    }
)


def default_py_config_path() -> Path:
    """Resolve the default JSON drop location.

    Precedence:

    1. ``$CORLINMAN_DATA_DIR/py-config.json``
    2. ``~/.corlinman/py-config.json``
    3. ``/tmp/corlinman-py-config.json`` (container-friendly fallback)
    """
    data_dir = os.environ.get("CORLINMAN_DATA_DIR")
    if data_dir:
        return Path(data_dir) / DEFAULT_PY_CONFIG_FILENAME
    home = Path.home() if _has_home() else None
    if home is not None:
        return home / ".corlinman" / DEFAULT_PY_CONFIG_FILENAME
    return Path("/tmp/corlinman-py-config.json")


def _has_home() -> bool:
    try:
        Path.home()
        return True
    except (RuntimeError, OSError):
        return False


# ---------------------------------------------------------------------------
# Renderer
# ---------------------------------------------------------------------------


def render_py_config(cfg: Any) -> dict[str, Any]:
    """Render the in-process config object as the Python JSON shape.

    Duck-typed against the Rust ``Config``:

    * ``cfg.providers`` — iterable of ``(name, entry)`` *or* a mapping
      ``{name: entry}``. Each ``entry`` exposes ``kind`` / ``api_key`` /
      ``base_url`` / ``enabled`` / ``params``.
    * ``cfg.models.aliases`` — mapping ``{alias_name: entry}``. ``entry``
      is either a full-form spec (``provider`` + ``model``) or a
      shorthand string. Shorthands are dropped — the Python legacy-prefix
      fallback handles them.
    * ``cfg.embedding`` — optional ``EmbeddingConfig`` with
      ``provider`` / ``model`` / ``dimension`` / ``enabled`` / ``params``.

    Invariants kept from the Rust impl:

    * Providers without a resolvable ``kind`` are dropped.
    * ``api_key`` is resolved via :func:`_resolve_secret`. Unresolved
      secrets render as ``None`` (Python ``ProviderSpec`` treats this as
      "no auth", same as the Rust side does for ``Option<String>``).
    * Shorthand / provider-less aliases are omitted from the output.
    """
    providers_in = _iter_providers(cfg)
    providers: list[dict[str, Any]] = []
    for name, entry in providers_in:
        kind = _kind_for(name, entry)
        if kind is None:
            continue
        providers.append(
            {
                "name": str(name),
                "kind": kind,
                "api_key": _resolve_secret(_attr(entry, "api_key", None)),
                "base_url": _attr(entry, "base_url", None),
                "enabled": bool(_attr(entry, "enabled", True)),
                "params": _params_to_json(_attr(entry, "params", {})),
            }
        )

    aliases_out: dict[str, Any] = {}
    for alias_name, alias_entry in _iter_aliases(cfg).items():
        rendered = _render_alias(alias_entry)
        if rendered is None:
            continue
        aliases_out[str(alias_name)] = rendered

    embedding = _render_embedding(_attr(cfg, "embedding", None))

    return {
        "providers": providers,
        "aliases": aliases_out,
        "embedding": embedding,
    }


def _iter_providers(cfg: Any) -> Iterable[tuple[str, Any]]:
    src = _attr(cfg, "providers", None)
    if src is None:
        return []
    if hasattr(src, "items") and callable(src.items):
        return list(src.items())
    if isinstance(src, Mapping):
        return list(src.items())
    # Try iterable of pairs.
    try:
        return [(n, e) for n, e in src]  # noqa: C416 — explicit unpack
    except TypeError:
        return []


def _iter_aliases(cfg: Any) -> Mapping[str, Any]:
    models = _attr(cfg, "models", None)
    aliases = _attr(models, "aliases", None) if models is not None else None
    if aliases is None:
        return {}
    if isinstance(aliases, Mapping):
        return aliases
    if hasattr(aliases, "items") and callable(aliases.items):
        return dict(aliases.items())
    return {}


def _render_alias(entry: Any) -> dict[str, Any] | None:
    # Shorthand: bare string.
    if isinstance(entry, str):
        return None
    provider = _attr(entry, "provider", None)
    model = _attr(entry, "model", None)
    # Provider-less full-form alias — same treatment as shorthand: omit.
    if not provider or not model:
        return None
    return {
        "provider": str(provider),
        "model": str(model),
        "params": _params_to_json(_attr(entry, "params", {})),
    }


def _render_embedding(emb: Any) -> dict[str, Any] | None:
    if emb is None:
        return None
    return {
        "provider": str(_attr(emb, "provider", "") or ""),
        "model": str(_attr(emb, "model", "") or ""),
        "dimension": int(_attr(emb, "dimension", 0) or 0),
        "enabled": bool(_attr(emb, "enabled", True)),
        "params": _params_to_json(_attr(emb, "params", {})),
    }


def _kind_for(name: str, entry: Any) -> str | None:
    explicit = _attr(entry, "kind", None)
    if explicit:
        kind = str(explicit)
        if kind in KNOWN_PROVIDER_KINDS:
            return kind
        # Unknown explicit kind — drop, matches the Rust ``None``-branch
        # behaviour.
        return None
    # Fallback: infer from the slot name. Mirrors
    # ``corlinman_core::config::Providers::kind_for`` for the first-party
    # names. Anything unrecognised drops out of the JSON.
    lowered = str(name).lower()
    if lowered in KNOWN_PROVIDER_KINDS:
        return lowered
    return None


def _resolve_secret(secret: Any) -> str | None:
    """Resolve a ``SecretRef``-shaped value.

    Accepts:

    * ``None`` → ``None``
    * a bare ``str`` → returned as-is (treated as a literal)
    * an object with ``env`` attr → ``os.environ.get(env)``
    * an object with ``value`` attr → ``str(value)``
    * a dict ``{"env": "..."}`` / ``{"value": "..."}`` — same shape as
      the Rust ``SecretRef`` tagged-enum serialization.
    """
    if secret is None:
        return None
    if isinstance(secret, str):
        return secret
    if isinstance(secret, Mapping):
        if "env" in secret:
            return os.environ.get(str(secret["env"]))
        if "value" in secret:
            value = secret["value"]
            return None if value is None else str(value)
        return None
    env = getattr(secret, "env", None)
    if env is not None:
        return os.environ.get(str(env))
    value = getattr(secret, "value", None)
    if value is not None:
        return str(value)
    return None


def _params_to_json(params: Any) -> dict[str, Any]:
    if params is None:
        return {}
    if isinstance(params, Mapping):
        return {str(k): _jsonable(v) for k, v in params.items()}
    if hasattr(params, "items") and callable(params.items):
        return {str(k): _jsonable(v) for k, v in params.items()}
    return {}


def _jsonable(value: Any) -> Any:
    """Pass-through coercion for ``serde_json::Value`` analogues.

    Pydantic models, dataclasses, and plain dict/list/str/int/float/bool
    are all already JSON-safe; anything else falls back to ``str()`` so
    the renderer can't 500 on an unexpected param value.
    """
    if isinstance(value, (str, int, float, bool, type(None))):
        return value
    if isinstance(value, Mapping):
        return {str(k): _jsonable(v) for k, v in value.items()}
    if isinstance(value, (list, tuple)):
        return [_jsonable(v) for v in value]
    if hasattr(value, "model_dump") and callable(value.model_dump):
        return value.model_dump()
    return str(value)


def _attr(obj: Any, name: str, default: Any) -> Any:
    """Mapping-aware ``getattr`` — looks up ``name`` on attributes and
    ``Mapping`` keys, so dicts and Pydantic models both work."""
    if obj is None:
        return default
    if isinstance(obj, Mapping):
        return obj.get(name, default)
    return getattr(obj, name, default)


# ---------------------------------------------------------------------------
# Writers
# ---------------------------------------------------------------------------


def write_py_config_sync(cfg: Any, path: Path | str) -> None:
    """Synchronously render + atomically write the JSON drop.

    Atomicity: write to a sibling ``<path>.new``, then ``os.rename`` —
    matches the Rust ``write_py_config_sync`` byte-for-byte so the reader
    side (``corlinman_server.main._ReloadingProviderResolver``) sees a
    fully-formed file on every mtime bump.
    """
    target = Path(path)
    payload = render_py_config(cfg)
    body = json.dumps(payload, indent=2, sort_keys=False) + "\n"
    if target.parent and not target.parent.exists():
        target.parent.mkdir(parents=True, exist_ok=True)
    # ``tempfile.NamedTemporaryFile`` keeps us safe from a half-written
    # file if the process dies mid-write; ``os.replace`` is atomic on
    # the same filesystem.
    fd, tmp = tempfile.mkstemp(
        prefix=target.name + ".",
        suffix=".new",
        dir=str(target.parent) if target.parent.as_posix() else ".",
    )
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as handle:
            handle.write(body)
        os.replace(tmp, target)
    except BaseException:
        # Clean up the temp file on any failure; the rename window is
        # microscopic but we don't want to leak ``.new`` shards.
        try:
            os.unlink(tmp)
        except OSError:
            pass
        raise


async def write_py_config(cfg: Any, path: Path | str) -> None:
    """Async wrapper around :func:`write_py_config_sync`.

    The Rust version uses ``tokio::fs`` for the rename hop. We stay sync
    here — the JSON payload is small (single-digit kB at most for a
    realistic provider set), and a serial sync write inside the async
    flow doesn't block the loop noticeably. Wrapping in
    :func:`asyncio.to_thread` would only buy us pretend-concurrency.
    """
    write_py_config_sync(cfg, path)


__all__ = [
    "DEFAULT_PY_CONFIG_FILENAME",
    "ENV_PY_CONFIG",
    "KNOWN_PROVIDER_KINDS",
    "default_py_config_path",
    "render_py_config",
    "write_py_config",
    "write_py_config_sync",
]
